buckets/redis/redis_keys.go

```diff
@@ -0,0 +1,136 @@
+// bucket names that are known to exist in Redis, based on the results of calling `existingBuckets()`, and `cfg` is the
+// map of NamespaceConfigs containing currently configured namespaces and buckets.
+func deleteUnknown(r *redis.Client, existing map[string][]string, cfg map[string]*qsc.NamespaceConfig) error {
+	redisKeysToDelete := findKeysToDelete(existing, cfg)
+
+	return chunkKeys(redisKeysToDelete, func (keys []string) error {
+		numDel, err := r.Del(keys...).Result()
+		if err != nil {
+			return err
+		}
+
+		logging.Printf("Purged %v unused keys from Redis", numDel)
+		return nil
+	}, REDIS_DEL_CHUNK_SIZE)
+}
+
+// findKeysToDelete inspects a map of namespace to bucket names (the starting collection of buckets), and a
+// NamespaceConfig. All buckets in the former, that do not exist in the latter, are converted to their Redis keys and
+// added to the slice that is returned. Note that buckets that aren't statically defined in a namespace, but in a
+// namespace that allows dynamic buckets, aren't considered candidates for removal. Global/Default buckets are never
+// considered for removal.
+func findKeysToDelete(existing map[string][]string, cfg map[string]*qsc.NamespaceConfig) []string {
+	redisKeysToDelete := make([]string, 0)
+
+	for nsName, bucketNames := range existing {
+		if nsName == config.GlobalNamespace {
+			continue
+		}
+
+		ns, exists := cfg[nsName]
+		for _, bucketName := range bucketNames {
+			if !exists {
+				// The entire namespace doesn't exist.
+				redisKeysToDelete = appendRedisKeys(redisKeysToDelete, nsName, bucketName)
+			} else if ns.DynamicBucketTemplate == nil {
+				// No dynamic buckets in this namespace. Check all buckets.
+				_, bucketExists := ns.Buckets[bucketName]
+				if !bucketExists && bucketName != config.DefaultBucketName {
+					// The bucket doesn't exist.
+					redisKeysToDelete = appendRedisKeys(redisKeysToDelete, nsName, bucketName)
+				}
+			}
+		}
+	}
+
+	return redisKeysToDelete
+}
+
+// appendRedisKeys creates all keys we expect to see for a given namespace/bucket name combo, and adds these keys to a
+// slice of keys passed in. Similar to Go's builtin `append()` function, you should always assign the return value of
+// this function to the slice variable you intend to use.
+func appendRedisKeys(keys []string, nsName, bName string) []string {
+	keys = append(keys, toRedisKey(nsName, bName, tokensNextAvblNanosSuffix))
+	return append(keys, toRedisKey(nsName, bName, accumulatedTokensSuffix))
+}
+
+// chunkKeys breaks up a slice of keys passed in into a smaller chunks, according to `chunkSize`, and invokes `callback`
+// with each chunk as a parameter. If the callback function returns an error, this function exists immediately,
+// propagating that error.
+func chunkKeys(keys []string, callback func([]string) error, chunkSize int) error {
+	// Chunk the keys to be deleted.
+	loops := len(keys) / chunkSize
```

Comment:
Sorry, ignore me - it's late and I'm tired :)

You're right, there's even a flaw in the test, and that's why it seemed OK.

Fixing.